[core] Implement utils class to setup and cleanup cgroup folder

dentiny opened the pull request:
This PR does a few things:
- It provides a class `CgroupV2Setup` which setups cgroupv2 at construction and deletes application cgroup at destruction;
- Raylet is able to add internal processes into the cgroup;
- Raylet is able to add application processes and their resource spec into the cgroup.

TODO:
- Integrate cgroupv2 setup class with raylet;
- Implement node-wise cgroup folder cleanup in raylet;
- Set resource reservation for internal processes.

dentiny committed:
utils to cleanup cgroup folder

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

edoakes commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -51,6 +55,9 @@ namespace {
 std::string cgroup_v2_app_folder;
 std::string cgroup_v2_system_folder;
 
+// Cgroup folder for the current ray node.
+std::string cgroup_v2_folder;
+
```
hm why have these as global variables rather than inside a class definition? the natural interface I was expecting is something like:
```
manager = CgroupManager("/path/to/node_cgroup") # manages everything under the path
manager.MoveProcesses(...)
manager.Cleanup()
```

edoakes replied:
this makes it very easy to swap out the fake implementation for testing and to configure the behavior and subpaths of the manager without mucking around in global variables

dentiny replied:
At the beginning when I made the implementation (two months ago), didn't expect a fake impl and unified interface.

edoakes commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -73,14 +80,55 @@ bool MoveProcsInSystemCgroup() {
   return out_file.good();
 }
 
+// This function is the reverse operation for [MoveProcsInSystemCgroup], which moves all
+// processes from raylet system cgroup into root cgroup.
+//
+// Return whether the PIDs move successfully.
+bool MoveProcsFromSystemCgroup() {
```
this function looks overfit to the specific functionality. prefer something like:
`MoveProcesses(const std::string dest)`

where `dest` is either a string identifier or maybe even an enum that enumerates the set of cgroups that we care to manage

dentiny replied:
> this function looks overfit to the specific functionality.

That's why it's placed in anonymous namespace in impl file, rather than a public header file.

edoakes commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -73,14 +80,55 @@ bool MoveProcsInSystemCgroup() {
   return out_file.good();
 }
 
+// Kill all processes inside of the cgroup.
+// If any operation fails, the error is logged and continue.
+void KillProcsUnderCgroupFolder(const std::string &folder) {
```
this interface should closely match the "MoveProcesses" functionality above

edoakes commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -73,14 +80,55 @@ bool MoveProcsInSystemCgroup() {
+  }
+
+  for (int pid : dangling_pids) {
+    if (kill(pid, SIGKILL) == -1) {
```
should we SIGTERM with a grace period first?

if this always immediately SIGKILLs, please document it in the header comment (not expected behavior)

dentiny replied:
The function is expected to invoke only at raylet destruction, at which point we expect all worker processes to complete and exit. But it's a good call to clearly document the behavior.

dentiny replied:
Some system uses SIGTERM because they have sigterm handler and a drain mechanism, which rejects all new incoming requests and gracefully shutdown, but I don't think it's something applicable to ray worker process.

edoakes commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,6 +227,38 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
   return true;
 }
 
+void CleanupCgroupForNode(const std::string &node_id) {
```
in line with the suggestions above, each raylet manages only a single node so there's really no reason we should need to pass the `node_id` in beyond initializing the cgroup manager obj

edoakes replied:
I would actually consider it an abstraction leak that the cgroup manager knows anything about a "node" at all

dentiny replied:
I wrap all related functions into a class, so no need to pass arguments any more.

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
cleanup cgroup

Signed-off-by: dentiny <dentinyhao@gmail.com>

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -140,31 +185,35 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
-  cgroup_v2_app_folder = absl::StrFormat("%s/ray_application_%s", cgroup_folder, node_id);
-  cgroup_v2_system_folder = absl::StrFormat("%s/ray_system_%s", cgroup_folder, node_id);
+  cgroup_v2_app_folder_ =
+      absl::StrFormat("%s/ray_application_%s", cgroup_v2_folder_, node_id);
```
This looks different from the [README](https://github.com/ray-project/ray/blob/master/src/ray/common/cgroup/README.md):

```
 /sys/fs/cgroup/ray_node_<node_id>
        /                       \
.../internal           .../application
                        /            \
                .../default  .../<task_id>_<attempt_id> (*N)
```

This looks like `/sys/fs/cgroup/ray_node_<id>/ray_application_<node_id>`

Why repeat the node_id twice instead of:
`/sys/fs/cgroup/ray_node_<id>/application`

dentiny replied:
fixed

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -140,31 +185,35 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
-  cgroup_v2_system_folder = absl::StrFormat("%s/ray_system_%s", cgroup_folder, node_id);
+  cgroup_v2_app_folder_ =
+      absl::StrFormat("%s/ray_application_%s", cgroup_v2_folder_, node_id);
+  cgroup_v2_system_folder_ =
```
Same as above. As per the README:

` /sys/fs/cgroup/ray_node_<node_id>/internal`

In this PR:

` /sys/fs/cgroup/ray_node_<node_id>/ray_system_<node_id>`

I prefer the format from the README

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -63,7 +67,7 @@ constexpr int kCgroupFilePerm = 0600;
 // Return whether the PIDs move successfully.
-bool MoveProcsInSystemCgroup() {
+bool MoveProcsInSystemCgroup(const std::string &cgroup_v2_system_folder) {
   std::ifstream in_file(kRtootCgroupProcs.data());
```
Isn't this going to move _all_ processes from the root cgroup to the ray_system cgroup?

dentiny replied:
Yes?

dentiny replied:
> Oh, is this for the Docker implementation only? Can you update the PR description to clarify.

Done.

> based on our conversation today, would it make more sense to keep all processes except for internal processes in the application cgroup?

What's the reason?

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -140,31 +185,35 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
     return false;
   }
   if (!EnableCgroupSubtreeControl(kRootCgroupSubtreeControl.data())) {
     return false;
```
At this point, we've created cgroups and migrated processes into the system cgroup. Should we still just return false or should we throw a fatal error b/c this should never happen.

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -140,31 +185,35 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
 
   // Setup application cgroup.
-  ret_code = mkdir(cgroup_v2_app_folder.data(), kCgroupFilePerm);
+  ret_code = mkdir(cgroup_v2_app_folder_.data(), kCgroupFilePerm);
```
Same question as above for error handling

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -122,13 +167,13 @@ bool IsCgroupV2MountedAsRw() {
+  cgroup_enabled_ = SetupCgroupsPreparation(node_id_);
+}
 
+bool CgroupSetup::SetupCgroupsPreparation(const std::string &node_id) {
```
IMO we should have two failure modes:

- Precondition checking (where we return false and don't use cgroups)
- Fatal errors if we've partially created cgroups and migrated some processes

What do you think?

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,10 +225,111 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+      ray::JoinPaths(cgroup_v2_app_folder_, "default");
+  const bool dir_succ = std::filesystem::create_directory(default_cgroup_folder);
+  if (!dir_succ) {
+    RAY_LOG(ERROR) << "Fails to create cgroup " << default_cgroup_folder;
```
s/Fails/Failed

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,10 +225,111 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+  }
+
+  auto delete_cgroup_folder = [folder = default_cgroup_folder]() {
+    RAY_CHECK(std::filesystem::remove(folder));  // Not expected to fail.
```
For my understanding, why is there a CHECK here?

dentiny replied:
My point is:
- Cgroup folder is managed by ray, at this point we've successfully created the cgroup so it should exist;
- Delete a cgroup directory only involves in-memory operation, so except for INTERNAL errors (aka, we wrote a bug) the operation shouldn't fail

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,10 +225,111 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+  std::ofstream out_file(cgroup_proc_file);
+  out_file << ctx.pid;
+  if (!out_file.good()) {
+    RAY_LOG(ERROR) << "Fails to add process " << ctx.pid << " with max memory "
```
I think this is a Fatal Error as well. If we've already setup some of the cgroups and we can't add an application process to the correct cgroup, it's an internal error.

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,10 +225,111 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+ScopedCgroupHandler CgroupSetup::ApplyCgroupForDefaultAppCgroup(
+    const AppProcCgroupMetadata &ctx) {
+  RAY_CHECK_EQ(ctx.max_memory, 0);  // Sanity check.
+  const std::string default_cgroup_folder =
```
Where is this folder created? Should it be in `SetupCgroupsPreparation`?

dentiny replied:
Yes that's correct, and that's my implementation?

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,10 +225,111 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+  std::ofstream out_file(default_cgroup_proc_file);
+  out_file << ctx.pid;
+  if (!out_file.good()) {
+    RAY_LOG(ERROR) << "Fails to add process " << ctx.pid << " with max memory "
```
Probably also a Fatal error.

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,10 +225,111 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+}
+
+// TODO(hjiang): Make a few constants for cgroup.
+ScopedCgroupHandler CgroupSetup::ApplyCgroupForDefaultAppCgroup(
```
Why return a ScopedCgroupHandler if there's never any cleanup?

dentiny replied:
Same question answered above.

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny commented on src/ray/common/cgroup/base_cgroup_setup.h:
```diff
@@ -46,13 +46,6 @@ class BaseCgroupSetup {
   // Apply cgroup context, which adds the process id into the corresponding cgroup.
   virtual ScopedCgroupHandler ApplyCgroupContext(const AppProcCgroupMetadata &ctx) = 0;
-
- protected:
```
It's not used in the production implementation.

dentiny replied:
I don't understand, I already remove it?

dentiny replied:
That was my mistake.

dentiny commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,10 +225,111 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+  }
+}
+
+ScopedCgroupHandler CgroupSetup::AddSystemProcess(pid_t pid) {
```
We need a unified return value has sth to cleanup && nothing to cleanup.

dentiny replied:
my bad, mis-understood, remove token to return

dentiny commented on src/ray/common/cgroup/cgroup_setup.h:
```diff
@@ -28,19 +30,65 @@ bool IsCgroupV2MountedAsRw();
 
-// Get folder name for application cgroup v2 for current raylet instance.
-const std::string &GetCgroupV2AppFolder();
+  [[maybe_unused]] bool cgroup_enabled_ = true;
```
I'm worried for windows platform, compiler would warn and error for unused data member.

dentiny commented on src/ray/common/cgroup/cgroup_setup.h:
```diff
@@ -28,19 +30,65 @@ bool IsCgroupV2MountedAsRw();
-const std::string &GetCgroupV2AppFolder();
+  [[maybe_unused]] bool cgroup_enabled_ = true;
+  [[maybe_unused]] const std::string node_id_;
+  // Root folder for cgroup v2 for the current raylet instance.
```
I think the comment is confusion, updated

dentiny commented on src/ray/common/cgroup/cgroup_setup.h:
```diff
@@ -16,6 +16,8 @@
+
 namespace ray {
 
 namespace internal {
```
Because it's not related to environment (i.e. docker / VM / BM) and could be unit tested.

dentiny replied:
I use namespace `internal` to indicate it's not supposed to be used outside.

dentiny commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,10 +225,111 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+  const std::string cgroup_folder = ray::JoinPaths(cgroup_v2_app_folder_, ctx.id);
+  // Create cgroup folder.
+  const std::string default_cgroup_folder =
+      ray::JoinPaths(cgroup_v2_app_folder_, "default");
```
Default application is not related to task attempt.

israbbani replied:
I think the `ApplyCgroupForIndividualAppCgroup` is for milestone 2 actor and task resource management, right? It shouldn't be under the default application cgroup

dentiny committed:
address comments 1st round

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -73,14 +77,55 @@ bool MoveProcsInSystemCgroup() {
+}
+
 // Return whether cgroup control writes successfully.
 bool EnableCgroupSubtreeControl(const char *subtree_control_path) {
```
I plan to enable both memory and CPU in the future here. I leave a TODO for CPU resource :)

dentiny committed:
address comment 2nd round

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -176,10 +225,111 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+  return ScopedCgroupHandler{};
+}
+
+ScopedCgroupHandler CgroupSetup::ApplyCgroupForIndividualAppCgroup(
```
Discussed offline, at the end of the day we will have an extra flag to indicate whether proceed or exit if failure.
Also add a TODO to record current behavior:
- For INTERNAL errors, log and exit directly with RAY_LOG(ERROR);
- For non-internal errors, log and return immediately.

dentiny committed:
fix application cgroup folder

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny committed:
leverage once token

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
no need for handle for system proc

Signed-off-by: dentiny <dentinyhao@gmail.com>

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -37,50 +49,77 @@ bool IsCgroupV2MountedAsRw() { return false; }
 
+// Kill all processes inside of the cgroup.
+// If any operation fails, the error is logged and continue.
+void KillProcsUnderCgroupFolder(const std::string &folder) {
```
Reading the [manual](https://docs.kernel.org/admin-guide/cgroup-v2.html) today, I saw a cgroup.kill interface. Can we use that instead of manually calling SIGKILL?

dentiny replied:
Nice!

israbbani commented on src/ray/common/cgroup/cgroup_setup.h:
```diff
@@ -28,19 +34,61 @@ bool IsCgroupV2MountedAsRw();
+  // - Existing operating system processes and system components will be placed under
+  // system cgroup. For more details, see
+  // https://github.com/ray-project/ray/blob/master/src/ray/common/cgroup/README.md
+  explicit CgroupSetup(std::string node_id);
```
const std::string&?

israbbani reviewed (commented):
A few more things (in addition to the comments)

- s/system/internal for all cgroup related stuff to be consistent with the final API

dentiny committed:
address comment 3

Signed-off-by: dentiny <dentinyhao@gmail.com>

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -37,50 +49,56 @@ bool IsCgroupV2MountedAsRw() { return false; }
+  const std::string kill_proc_file = absl::StrFormat("%s/cgroup.kill", cgroup_folder);
+  std::ofstream f{kill_proc_file};
+  f << "1";
+  RAY_CHECK(f.good()) << "Fails to kill all processes under the cgroup";
```
Would logging the cgroup_folder provide useful information to someone debugging this failure?

dentiny replied:
makes sense

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -122,64 +140,121 @@ bool IsCgroupV2MountedAsRw() {
+  RAY_CHECK(internal::CanCurrenUserWriteCgroupV2())
+      << "Current user doesn't have the permission to update cgroup v2.";
+
+  RAY_CHECK(internal::IsCgroupV2MountedAsRw())
```
I think it makes more sense to do the checks in the opposite order. 

1. check to see if cgroupv2 is mounted
2. if it is mounted, check to see if the current user can read/write to the cgroup 

There are two more potential issues with the implementation of IsCgroupV2MountedAsRw (I can't comment on it directly because that code wasn't committed as part of this PR):

1. Having cgroupv2 mounted is not enough. We need to make sure that cgroupv1 is not mounted and we only support unified mode (not hybrid)
2. We should check to make sure cgroupv2 is mounted at the path we expect /sys/fs/cgroup or we can take the path from the mounts file  

It'll also be good to check to see if the controllers we want to use (memory, cpu) are available and enabled in the root (as presented to us) cgroup.

israbbani reviewed (commented):
A few more issues:
* Potential bug with using ofstream.good() to see if writes to virtual files succeeded
* Potential hardening of detecting if cgroupv2 is enabled (should check for unified mode)

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -37,50 +49,77 @@ bool IsCgroupV2MountedAsRw() { return false; }
+  RAY_CHECK(out_file.good()) << "Failed to open cgroup file " << subtree_control_path;
+  // Able to add memory constraint to the system cgroup.
+  out_file << "+memory";
   return out_file.good();
```
Debugging this further, I think we want to use synchronous I/O to guarantee the buffer is flushed before we do the check:
```
(base) ubuntu@devbox:~/investigations/cgroup$ sudo su
root@devbox:/home/ubuntu/investigations/cgroup# strace ./cgroup_test.o 2>&1 | tail -10
brk(NULL)                               = 0x560e2d0d3000
brk(0x560e2d0f4000)                     = 0x560e2d0f4000
openat(AT_FDCWD, "/sys/fs/cgroup/Example/cgroup.subtree_control", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 3
fstat(1, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
**write(1, "out_file.good() = 1\n", 20out_file.good() = 1
)   = 20**
**write(3, "complete_garbage", 16)        = -1 EINVAL (Invalid argument)**
close(3)                                = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

dentiny replied:
Thanks for the investigation!

israbbani commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -37,50 +49,56 @@ bool IsCgroupV2MountedAsRw() { return false; }
-  int pid = 0;
+// Move all pids under [from] to [to].
+bool MoveProcsBetweenCgroups(const std::string &from, const std::string &to) {
+  std::ifstream in_file(to.data());
```
Should this be the other way around? 

`std::ifstream in_file(from.data())`
`std::ofstream out_file(to.data())` 

Also, does ofstream open with O_TRUNC by default. On a virtual file system, will it delete all existing pids in the file?

dentiny commented:
> Potential hardening of detecting if cgroupv2 is enabled (should check for unified mode)

Addressed in a separate PR: https://github.com/ray-project/ray/pull/51141

dentiny committed:
fix cgroup file flush

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -140,46 +179,124 @@ bool SetupCgroupsPreparation(const std::string &node_id) {
+    return;
   }
+  std::ofstream out_file(cgroup_v2_system_folder_);
+  // Able to add memory constraint to the system cgroup.
```
The code doesn't exist any more, but looks like the place we add memory control to subtree control.
So to enable memory control for subcgroups, we need to add `+memory` to cgroup subcontrol file.

dentiny committed:
address comment

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
use permission bits

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
make sure flush and open with append

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
cleanup unused functions

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
extract constant

Signed-off-by: dentiny <dentinyhao@gmail.com>

MortalHappiness committed:
[Test][KubeRay] Add doctest for RayCluster Quickstart doc (#51249)

Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

Commit:
[ci] Enable Cgroup support in CI for core (#51454)

Enable CI environment with writeable cgroupv2 in CI. 

To use this environment for testing, you need to use the
`--privileged-container` flag in your test definitions.
To enable bazel read/write to sys/fs/cgroup in your tests, you need to
use the `--build-type cgroup`

---------

Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

abrarsheikh committed:
refactor replica _handle_errors_and_metrics (#51644)

Signed-off-by: Abrar Sheikh <abrar@anyscale.com>

comaniac committed:
[ray.llm] Refactor model download utilities (#51604)

Signed-off-by: Cody Yu <hao.yu.cody@gmail.com>

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

akyang-anyscale committed:
[serve] Remove RAY_SERVE_ENABLE_QUEUE_LENGTH_CACHE flag (#51649)

## Why are these changes needed?

This PR removes a feature flag that controls whether the proxy should
use cached replica queue length values for routing. The FF was
[introduced](https://github.com/ray-project/ray/pull/42943) over a year
ago as a way for users to quickly switch back to the previous
implementation. It has been enabled by default for [over a
year](https://github.com/ray-project/ray/pull/43169) now and works as
expected, so let's remove it. Consequently, this PR also removes
`RAY_SERVE_ENABLE_STRICT_MAX_ONGOING_REQUESTS`, as it is always enabled
if `RAY_SERVE_ENABLE_QUEUE_LENGTH_CACHE` is enabled.

Signed-off-by: akyang-anyscale <alexyang@anyscale.com>

dentiny committed:
fix build

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
use constants

Signed-off-by: dentiny <dentinyhao@gmail.com>

leibovitzgil committed:
[data] Fix Databricks host URL handling in Ray Data (#49926)

<!-- Thank you for your contribution! Please review
https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before
opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR.
If you don't have the access to it, we will shortly find a reviewer and
assign them to your PR. -->

## Why are these changes needed?

The current Databricks integration in Ray Data requires providing the
Databricks host URL without the "https://" prefix. However, this creates
compatibility issues when using Ray Data alongside MLflow, as MLflow's
Databricks integration (which uses the same DATABRICKS_HOST environment
variable) expects the URL to include the "https://" prefix.

## Related issue number

Closes #49925 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

---------

Signed-off-by: Gil Leibovitz <gil.leibovitz@doubleverify.com>
Signed-off-by: Richard Liaw <rliaw@berkeley.edu>
Co-authored-by: Gil Leibovitz <gil.leibovitz@doubleverify.com>
Co-authored-by: Richard Liaw <rliaw@berkeley.edu>

dentiny committed:
remove milestone-2 code

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
doc on internal process cgroup addition

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
doc for error handling

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
return status instead of bool

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix add pid into file

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
macro for cgroup operations

Signed-off-by: dentiny <dentinyhao@gmail.com>

srinathk10 committed:
[data] Update repartition on target_num_rows_per_block documentation (#51433)

<!-- Thank you for your contribution! Please review
https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before
opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR.
If you don't have the access to it, we will shortly find a reviewer and
assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this
solves. -->

Update repartition on target_num_rows_per_block:
When `target_num_rows_per_block` is set, it only repartitions Dataset
blocks that are larger than `target_num_rows_per_block`.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

---------

Signed-off-by: Srinath Krishnamachari <srinath.krishnamachari@anyscale.com>

khluu committed:
[deps] Use UV to compile LLM dependencies (#51323)

uv much much faster

also adds tests to enforce dependency integrity.

Signed-off-by: kevin <kevin@anyscale.com>

dentiny committed:
apply cgroup status macros

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix windows

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
consider cgroup types

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
initial setup test

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
initial setup test

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
propagate with filename and line no

Signed-off-by: dentiny <dentinyhao@gmail.com>

aslonnie committed:
[ci] add an always tag for cond testing (#51662)

so that "lint" tag can be used for just lints, and will not be abused to
be used as "always"

Signed-off-by: Lonnie Liu <lonnie@anyscale.com>

dentiny committed:
fix initialization order

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix already exists

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
folder name in error message

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
delete application folder

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix unit test

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
compile option

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
use macros

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
check internal pid addition

Signed-off-by: dentiny <dentinyhao@gmail.com>

kevin85421 committed:
[core] Correct the wording in the OnNodeDead logs to avoid confusion (#51668)

Signed-off-by: Kai-Hsun Chen <kaihsun@anyscale.com>

Jay-ju committed:
[data] fix lance ut failed (#51421)

<!-- Thank you for your contribution! Please review
https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before
opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR.
If you don't have the access to it, we will shortly find a reviewer and
assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this
solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Signed-off-by: jukejian <jukejian@bytedance.com>

tespent committed:
[Data] fix RandomAccessDataset.multiget returning unexpected values for missing keys (#44769)

<!-- Thank you for your contribution! Please review
https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before
opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR.
If you don't have the access to it, we will shortly find a reviewer and
assign them to your PR. -->

As stated in #44768, the current implementation of `multiget` based on
`np.searchsorted` does not check for missing keys. I added the required
checks and updated unit test for this case.

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this
solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->
Closes #44768 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Signed-off-by: Wu Yufei <wuyufei.2000@bytedance.com>

Drice1999 committed:
[Data] Support async callable classes in flat_map() (#51180)

<!-- Thank you for your contribution! Please review
https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before
opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR.
If you don't have the access to it, we will shortly find a reviewer and
assign them to your PR. -->

## Why are these changes needed?
This PR adds async generator support to `flat_map`.
The implementation is similar to how #46129 handled async callable
classes for map_batches(), changes include:
* Generalize the logic in `_generate_transform_fn_for_async_flat_map` so
it can process both batches and rows
* Add test case for async `flat_map`
<!-- Please give a short summary of the change and the problem this
solves. -->

## Related issue number
#50329 
<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

---------

Signed-off-by: Drice1999 <chenxh267@gmail.com>

dentiny committed:
[core] [easy] Mark cgroup tests exclusive (#51654)

cgroup tests access (read and write) system directory, so they shouldn't
be executed in parallel.
For example, `bazel test //src/ray/common/cgroup/tests:all` should run
tests one by one.

---------

Signed-off-by: dentiny <dentinyhao@gmail.com>

kevin85421 committed:
[core] Threaded actors get stuck forever if they receive two exit signals (#51582)

If a threaded actor receives exit signals twice as shown in the above
screenshot, it will execute
[task_receiver_->Stop()](https://github.com/ray-project/ray/blob/6bb9cef9257046ae31f78f6c52015a8ebf009f81/src/ray/core_worker/core_worker.cc#L1224)
twice. However, the second call to `task_receiver_->Stop()` will get
stuck forever when executing the
[releaser](https://github.com/ray-project/ray/blob/6bb9cef9257046ae31f78f6c52015a8ebf009f81/src/ray/core_worker/transport/concurrency_group_manager.cc#L135).

### Reproduction

```sh
ray start --head --include-dashboard=True --num-cpus=1
# https://gist.github.com/kevin85421/7a42ac3693537c2148fa554065bb5223
python3 test.py

# Some actors are still ALIVE. If all actors are DEAD, increase the number of actors.
ray list actors
```

---------

Signed-off-by: kaihsun <kaihsun@anyscale.com>

dentiny committed:
[core] Fix all variable shadowing for core worker (#51672)

This PR fixes all variable shadowing for core worker.

Variable shadowing is known to be error prune --- I personally just
spent 20 minute debugging an issue caused by it.
Need to address all issues before
https://github.com/ray-project/ray/pull/51669

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
disable env for cgroup test and fix file open

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix unit test

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
TODO

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
only cleanup application cgroup

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
friend

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix non linux

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
no check at dtor

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
extract test utils

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
Add test for application cgroup

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny committed:
no shadow for now

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
default app

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
delete subdirectory

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix compile

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix proc filepath

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix unit test, block wait

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix windows

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny commented on src/ray/common/cgroup/base_cgroup_setup.h:
```diff
@@ -42,17 +42,10 @@ class BaseCgroupSetup {
-
- protected:
-  // Remove the given system process [pid] from system cgroup.
-  virtual void CleanupSystemProcess(pid_t pid) = 0;
```
We don't need a util function to cleanup internal processes, because all of them are placed at the same cgroup folder, and gets removed from the PID file after they complete.

The only thing we need to do, is to delete the delete the internal sub-cgroup after raylet exits, using https://github.com/ray-project/ray/blob/master/src/ray/util/process_cleaner.h

dentiny committed:
fix unit test

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
join path

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
cpu as well

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix comment and rename

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny commented:
This PR should be ok to review, because failed CI test has nothing to do with my change.

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny committed:
fix comparison

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny committed:
use system instead of internal

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
suppress via unused

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny committed:
fix for compilation option

Signed-off-by: dentiny <dentinyhao@gmail.com>

israbbani commented on src/ray/common/cgroup/cgroup_setup.h:
```diff
@@ -33,21 +38,81 @@ Status CheckCgroupV2MountedRW(const std::string &directory);
+  ScopedCgroupHandler ApplyCgroupContext(const AppProcCgroupMetadata &ctx) override;
+
+ private:
+  struct Tag {};
+  // Constructor made for unit tests, which allows [CgroupSetup] to be created for
+  // multiple times in a process.
+  CgroupSetup(const std::string &directory, const std::string &node_id, Tag);
```
Can we call this something that conveys that it's only used for testing? 

TestDummy? TestTag?

dentiny replied:
Updated to `TestTag` since no other usage. 
Tag dispatch is a technique in C++.

israbbani commented on src/ray/common/cgroup/test/cgroup_v2_setup_test.cc:
```diff
@@ -0,0 +1,138 @@
+
+class Cgroupv2SetupTest : public ::testing::Test {
+ public:
+  Cgroupv2SetupTest()
+      : node_id_("node_id"),
+        node_cgroup_folder_("/sys/fs/cgroup/ray_node_node_id"),
+        internal_cgroup_folder_("/sys/fs/cgroup/ray_node_node_id/internal"),
+        internal_cgroup_proc_filepath_(
+            "/sys/fs/cgroup/ray_node_node_id/internal/cgroup.procs"),
+        app_cgroup_folder_("/sys/fs/cgroup/ray_node_node_id/ray_application"),
+        app_cgroup_proc_filepath_(
+            "/sys/fs/cgroup/ray_node_node_id/ray_application/default/cgroup.procs") {}
+  void TearDown() override {
+    // Check the node-wise subcgroup folder has been deleted.
+    std::error_code err_code;
+    bool exists = std::filesystem::exists(app_cgroup_folder_, err_code);
+    ASSERT_FALSE(err_code);
+    ASSERT_FALSE(exists);
+  }
+
+ protected:
+  const std::string node_id_;
+  const std::string node_cgroup_folder_;
+  const std::string internal_cgroup_folder_;
+  const std::string internal_cgroup_proc_filepath_;
+  const std::string app_cgroup_folder_;
+  const std::string app_cgroup_proc_filepath_;
+};
```
Using fixtures for unit tests can make tests harder to read and not self-contained. It's one of the only places in which I prefer verbose/repetitive code over code reuse. 

Since this fixture is pretty trivial, I would replace by adding the necessary code to each test to make it more readable.

dentiny replied:
The reason why I use text fixture (a class) is the constants and check logic are shared.

dentiny replied:
> Using fixtures for unit tests can make tests harder to read and not self-contained

Could you please explain why tests are not self-contained?

israbbani replied:
We synced up offline. I think we have different opinions on this but it's fine.

israbbani commented on src/ray/common/cgroup/test/cgroup_v2_setup_test.cc:
```diff
@@ -0,0 +1,138 @@
+}
+#else
+
+class Cgroupv2SetupTest : public ::testing::Test {
```
Thank you for doing this. I think we need a better culture of C++ unit testing in the core codebase. I'd like to see more extensive testing for the unhappy path, but that can happen in another PR.

dentiny replied:
I think we've discussed briefly, a few unhappy paths on top of my head:
- Not mounted, or don't have permission (already tested in general testing);
- Mounted v1 (need special setup);
- Mounted not as rw mode (need special setup).

I understand for testing thoroughness it's good to have them, but we need to figure out a strategy.

dentiny commented on src/ray/common/cgroup/BUILD:
```diff
@@ -59,3 +57,14 @@ ray_cc_library(
+    deps = [
+        "//src/ray/common:status",
+        "@com_google_absl//absl/strings:str_format",
+        "@com_google_googletest//:gtest_prod",
```
Yes, it's a necessary dependency for `FRIEND_TEST` macro.

dentiny commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -27,66 +40,105 @@ Status CheckCgroupV2MountedRW(const std::string &directory) {
+//
+// TODO(hjiang): Currently only memory resource is considered, should consider CPU
+// resource as well.
+Status EnableCgroupSubtreeControl(const char *subtree_control_path) {
```
No strong opinion, change it back.

dentiny commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -27,66 +40,105 @@ Status CheckCgroupV2MountedRW(const std::string &directory) {
+  RAY_SCHECK_OK_CGROUP(out_file.good())
+      << "Failed to flush cgroup file " << subtree_control_path;
+
+  return Status::OK();
```
Yes, I quickly checked in my local
```sh
[~/ray] (hjiang/cleanup-cgroup) [gke_hjiang-proj-cgroup-experiment_us-west1_hjiang-cgroup-test]
ubuntu@hjiang-devbox-pg$ cat /sys/fs/cgroup/cgroup.controllers 
cpuset cpu io memory hugetlb pids rdma misc
(myenv) (myenv) 
[~/ray] (hjiang/cleanup-cgroup) [gke_hjiang-proj-cgroup-experiment_us-west1_hjiang-cgroup-test]
ubuntu@hjiang-devbox-pg$ echo "+freezer" | sudo tee /sys/fs/cgroup/ray/cgroup.subtree_control
+freezer
tee: /sys/fs/cgroup/ray/cgroup.subtree_control: Invalid argument
```
Write operation will fail in this case.

dentiny commented on src/ray/common/cgroup/cgroup_setup.cc:
```diff
@@ -27,66 +40,105 @@ Status CheckCgroupV2MountedRW(const std::string &directory) {
+StatusOr<bool> IsRootCgroup(const std::string &directory) {
+  const std::string cgroup_type_filepath = ray::JoinPaths(directory, kCgroupTypeFilename);
+  std::error_code err_code;
+  bool exists = std::filesystem::exists(cgroup_type_filepath, err_code);
```
Ah good catch!

dentiny commented on src/ray/common/cgroup/test/cgroup_v2_setup_test.cc:
```diff
@@ -0,0 +1,138 @@
+  if (pid == 0) {
+    // Spawn a process running long enough, so it could be added into internal cgroup.
+    // It won't affect test runtime, because it will be killed later.
+    execlp("sleep", "sleep", "3600", nullptr);
```
You can use sleep in the child process, but gtest related things are forked as well, which means you runs other test cases, and print the testing results out in the stdout/stderr.

One related example I could give is: https://github.com/ray-project/ray/pull/50929

dentiny commented on src/ray/common/cgroup/test/cgroup_v2_setup_test.cc:
```diff
@@ -0,0 +1,138 @@
+//
+// Setup command:
+// sudo umount /sys/fs/cgroup/unified
+// sudo mount -t cgroup2 cgroup2 /sys/fs/cgroup/unified -o rw
```
This is the step I did in my ubuntu machine when I initially setup.

israbbani replied:
```
// Check if you have cgroupv2 mounted correctly.
// Your output should look like this to run tests correctly
// mount | grep cgroup
// cgroup2 on /sys/fs/cgroup type cgroup2 // (rw,nosuid,nodev,noexec,relatime,nsdelegate)
//
// To mount cgroupv2, follow this guide:
// https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/8/html/managing_monitoring_and_updating_the_kernel/using-cgroups-v2-to-control-distribution-of-cpu-time-for-applications_managing-monitoring-and-updating-the-kernel#mounting-cgroups-v2_using-cgroups-v2-to-control-distribution-of-cpu-time-for-applications
// TODO: Create a README file that can be linked here.

dentiny committed:
address comment

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny committed:
address comment

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
TODO for instruction

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
add link

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
sleep instead of exec

Signed-off-by: dentiny <dentinyhao@gmail.com>

israbbani reviewed (approved):
🚢

dentiny committed:
fix lint

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
temp workaround

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
no need TODO

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

israbbani commented:
We found a _real issue_ in CI w/ unit tests so that's a win for having to go through all this work 🚀 

Lets hold off on merging this PR while we address:
- Creating nested cgroups for application and system processes (because of the https://docs.kernel.org/admin-guide/cgroup-v2.html#no-internal-process-constraint) 
- Enabling cgroup.controllers on the top level cgroup that is passed to the raylet
- Checking `cgroup.controllers` to make sure memory and cpu controllers are available. I'm happy to do this check in another PR if you leave a TODO.

```bash
# precondition to check that memory and cpu controller are available
root@a04912fe501a:/# cat /sys/fs/cgroup/cgroup.controllers
cpuset cpu io memory hugetlb pids rdma misc

mkdir /sys/fs/cgroup/ray_node
mkdir /sys/fs/cgroup/ray_node/system
mkdir /sys/fs/cgroup/ray_node/system/system
mkdir /sys/fs/cgroup/ray_node/app
mkdir /sys/fs/cgroup/ray_node/app/app
for proc in `cat /sys/fs/cgroup/cgroup.procs`; 
  do echo $proc >> /sys/fs/cgroup/ray_node/system/system; done
echo "+memory" >> /sys/fs/cgroup/
echo "+memory" >> /sys/fs/cgroup/ray_node
echo "+memory" >> /sys/fs/cgroup/ray_node/system
```

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny committed:
Merge branch 'master' into hjiang/cleanup-cgroup

dentiny committed:
fix non root env proc movement

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
no data

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
no precheck

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix constant

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix status or

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
check write as well

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
check leaf node at failure

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
check root cgroup

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix logging

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
rename internal to system

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
fix source filename

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
check all pids

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
no flush for cgroup files

Signed-off-by: dentiny <dentinyhao@gmail.com>

dentiny committed:
cleanup

Signed-off-by: dentiny <dentinyhao@gmail.com>
